content/codes/components/enigma-path-svg.ts

```diff
@@ -0,0 +1,612 @@
+		tolamp.pts.push(makept(x1, y + y1 * dy));
+		g = y1 % 2 ? 33 : 53;
+		tolamp.pts.push(makept(x1 + g, y + y1 * dy));
+		
+		const stack = [
+			...rotorsin, refin, refint, ...rotorsout, tolamp
+		];
+		
+		if (anim)
+		{
+			animatePath(x, y, dy, gap, width, encodingpath, stack, animpath);
+		}
+		else
+		{	
+			showPath(x, y, dy, gap, width, encodingpath, stack, encpath);
+		}
+	}
+
+	function showPath(x: number, y: number, dy: number, gap: number, width: number, encodingpath: string[], 
+		stack: any[], layer: SVGGElement)
+	{
+		const showSegment = function(seg: { pts: any[]; colour: string; arrow: { x: number; y: number; }; 
+			forward: boolean; label: { letter: string; pt: { x: number; y: number; }; cls: string; }; })
+		{
+			layer.innerHTML += path(pathpts(seg.pts), 2, seg.colour, "none");
+			if (seg.arrow)
+				drawArrow(layer, seg.arrow.x, seg.arrow.y, seg.forward);
+			if (seg.label)
+				addText(layer, seg.label.letter, seg.label.pt.x, seg.label.pt.y, seg.label.cls);
+		}	
+	
+		layer.innerHTML = '';
+		while (stack.length)
+		{
+			showSegment(stack.pop());
+		}
+	
+		x += (machine.isUsingPlugboard() ? 4 : 3) * (width + gap) + gap / 2 + 8;
+		highlightKeys(layer, x, y, dy, gap, encodingpath[0], encodingpath[9]);
+	}
+
+	function animatePath(x: number, y: number, dy: number, gap: number, width: number, encodingpath: string[], 
+		stack: any[], layer: SVGGElement)
+	{
+		const key = encodingpath[0];
+		const inkeyelement = makekey(0, 0, key, incolour, null);
+		
+		let anim = '';
+		 
+		layer.innerHTML += '';
+		anim += "<g id='inkeyanim'><g id='animkey'>" + inkeyelement + '</g>';
+		var i = 0;
+		while (i < stack.length)
+		{
+			const s = stack[i];
+			const pathid = "p" + i;
+			layer.innerHTML += path(pathpts(s.pts), 2, s.colour, "none", pathid);
+			const lastpath = svgdoc.getElementById(pathid) as unknown as SVGPathElement;
+			const pathlength:number = lastpath.getTotalLength();
+			lastpath.setAttribute("stroke-dasharray", pathlength.toString());
+			lastpath.setAttribute("stroke-dashoffset", pathlength.toString());
+			var duration = (pathlength ? Math.round(1000 * pathlength / animationspeed) : 1) + "ms";
+			var p = 'path="M' + s.pts.map((pt: { x: string; y: string; }) => pt.x + "," + pt.y).join("L") + '"';
+			var begin = i ? "a" + (i - 1) + ".end" : "indefinite";
+			anim += '<animateMotion id="a' + i + '" begin="' + begin + '" ' + p + 
+				' dur="' + duration + '" repeatCount="1" fill="freeze" />';
+
+	    var pathanim = document.createElementNS(xmlns, "animate");
+			pathanim.setAttributeNS(null, 'attributeName', "stroke-dashoffset");
+			pathanim.setAttributeNS(null, 'begin', i ? begin : "a0.begin");
+			pathanim.setAttributeNS(null, 'from', pathlength.toString());
+			pathanim.setAttributeNS(null, 'to', "0");
+			pathanim.setAttributeNS(null, 'dur', duration);
+			pathanim.setAttributeNS(null, 'fill', "freeze");
+			pathanim.setAttributeNS(null, 'repeatCount', "1");
+			lastpath.appendChild(pathanim);
```

Comment:
Rather than manually creating SVG elements like this, just add them to a template, or use something like `$N('animate', {stroke-dashoffset: '0px'})`. In this example, I think it is MUCH easier to just use CSS transitions or the [Web Animations API](https://www.notion.so/mathigon/Engineering-5272e83217004731bca2b63d03e654ea#de150aa87c284e998f1f2ce2c08b6e8a), rather than SVG animations.